packages/paymasters/src/TokenPaymasterProvider.ts

```diff
@@ -0,0 +1,129 @@
+  PermitInterfaceDAIInstance,
+  PermitInterfaceEIP2612Instance
+} from '../types/truffle-contracts'
+import { RelayRequest } from '@opengsn/common/dist/EIP712/RelayRequest'
+import {
+  PERMIT_SIGHASH_DAI, PERMIT_SIGHASH_EIP2612,
+  PERMIT_SIGNATURE_DAI,
+  PERMIT_SIGNATURE_EIP2612,
+  signAndEncodeDaiPermit,
+  signAndEncodeEIP2612Permit
+} from './PermitPaymasterUtils'
+import { constants } from '@opengsn/common/dist/Constants'
+import { EIP712Domain, EIP712DomainType, EIP712DomainTypeWithoutVersion } from '@opengsn/common/dist/EIP712/TypedRequestData'
+import { TokenPaymasterInteractor } from './TokenPaymasterInteractor'
+
+export interface TokenPaymasterConfig extends GSNConfig {
+  tokenAddress: Address
+  permitAmount?: string | number | BN
+}
+
+export interface TokenPaymasterUnresolvedConstructorInput extends GSNUnresolvedConstructorInput {
+  config: Partial<TokenPaymasterConfig>
+}
+
+export class TokenPaymasterProvider extends RelayProvider {
+  config!: TokenPaymasterConfig
+  token!: PermitInterfaceDAIInstance | PermitInterfaceEIP2612Instance
+  permitSignature!: string
+  protected paymaster!: PermitERC20UniswapV3PaymasterInstance
+  readonly tokenPaymasterInteractor: TokenPaymasterInteractor
+
+  constructor (relayClient: RelayClient, provider: Web3ProviderBaseInterface) {
+    super(relayClient)
+    this.tokenPaymasterInteractor = new TokenPaymasterInteractor(provider)
+  }
+
+  static newProvider (input: TokenPaymasterUnresolvedConstructorInput): TokenPaymasterProvider {
+    if (input.config.maxPaymasterDataLength == null) {
+      input.config.maxPaymasterDataLength = 300
+    }
+    const provider = new TokenPaymasterProvider(new RelayClient(input), input.provider)
+    return provider
+  }
+
+  async init (): Promise<this> {
+    await super.init(true)
+    this.relayClient.dependencies.asyncPaymasterData = this._buildPaymasterData.bind(this)
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    this.paymaster = await this.tokenPaymasterInteractor._createPermitERC20UniswapV3Paymaster(this.relayClient.config.tokenPaymasterAddress!)
+    if (this.config.tokenAddress != null) {
+      await this.useToken(this.config.tokenAddress)
+    }
+    return this
+  }
+
+  async _buildPaymasterData (relayRequest: RelayRequest): Promise<PrefixedHexString> {
+    //  Optionally encode permit method,then concatenate token address
+    if (relayRequest.relayData.paymaster !== this.paymaster.address) {
+      throw new Error('Paymaster address mismatch')
+    }
+    const allowance = await this.token.allowance(relayRequest.request.from, relayRequest.relayData.paymaster)
+    let permitMethod = ''
+    if (allowance.eqn(0)) {
+      if (this.config.tokenPaymasterDomainSeparators == null) {
+        throw new Error('TokenPaymasterProvider not initialized. Call init() first')
+      }
+      const domainSeparator: EIP712Domain = this.config.tokenPaymasterDomainSeparators[this.token.address]
```

Comment:
`.toLowerCase` the addresses as they sometimes get cHecKsuMmeD when you least expect it

Reply:
Using checksummed address only